pkg/istio-agent/xds_proxy.go

```diff
@@ -449,10 +409,55 @@ func (p *XdsProxy) HandleUpstream(ctx context.Context, con *ProxyConnection, xds
 }
 
 func (p *XdsProxy) handleUpstreamRequest(con *ProxyConnection) {
+	// Handle downstream xds
+	initialRequestsSent := atomic.NewBool(false)
+	go func() {
+		for {
+			// From Envoy
+			req, err := con.downstream.Recv()
+			if err != nil {
+				select {
+				case con.downstreamError <- err:
+				case <-con.stopChan:
+				}
+				return
+			}
+
+			// forward to istiod
+			con.sendRequest(req)
+			if !initialRequestsSent.Load() && req.TypeUrl == v3.ListenerType {
+				// fire off an initial NDS request
+				if _, f := p.handlers[v3.NameTableType]; f {
+					con.sendRequest(&discovery.DiscoveryRequest{
+						TypeUrl: v3.NameTableType,
+					})
+				}
+				// fire off an initial PCDS request
+				if _, f := p.handlers[v3.ProxyConfigType]; f {
+					con.sendRequest(&discovery.DiscoveryRequest{
+						TypeUrl: v3.ProxyConfigType,
+					})
+				}
+				// Fire of a configured initial request, if there is one
+				p.connectedMutex.RLock()
+				initialRequest := p.initialRequest
+				if initialRequest != nil {
+					con.sendRequest(initialRequest)
+				}
+				p.connectedMutex.RUnlock()
+				initialRequestsSent.Store(true)
```

Comment:
I think there is a race condition that will lead to droppped requests:

```
a) con.SendRequest
a) load p.initialRequest, it is nil. Unlock connectedMutex
b) PersistRequest, store initialRequest and send on req chan
c) read from requestChan. initialRequestsSent=false, do nothing
a) `initialRequestsSent.Store(true)`
```

Letter == the goroutine.

I think we can just move this Store into the connectedMutex lock?

Reply:
good catch, i think move the initialRequestsSent.Store(true) before the `con.sendRequest(initialRequest)` should be ok.